src/nvim/runtime.c

```diff
@@ -332,40 +332,55 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
   return done;
 }
 
-static void push_path(RuntimeSearchPath *search_path, char *entry, bool after)
+static void push_path(RuntimeSearchPath *search_path, Map(String, handle_T) *rtp_used,
+                      char *entry, bool after)
 {
-  kv_push(*search_path, ((SearchPathItem){ entry, after }));
+  handle_T *h = map_ref(String, handle_T)(rtp_used, cstr_as_string((char *)entry), true);
+  if (*h == 0) {
+    (*h)++;
+    kv_push(*search_path, ((SearchPathItem){ entry, after }));
+  }
 }
 
-static void expand_pack_entry(RuntimeSearchPath *search_path, CharVec *after_path,
-                              char_u *pack_entry)
+static void expand_rtp_entry(RuntimeSearchPath *search_path, Map(String, handle_T) *rtp_used,
+                             char *entry, bool after)
 {
-  static char_u buf[MAXPATHL], buf2[MAXPATHL];
-  char *start_dir = "/pack/*/start/*";  // NOLINT
-  if (STRLEN(pack_entry) + STRLEN(start_dir) + 1 < MAXPATHL) {
-    xstrlcpy((char *)buf, (char *)pack_entry, MAXPATHL);
-    xstrlcpy((char *)buf2, (char *)pack_entry, MAXPATHL);
-    xstrlcat((char *)buf, start_dir, sizeof buf);
-    xstrlcat((char *)buf2, "/start/*", sizeof buf);  // NOLINT
-    int num_files;
-    char_u **files;
-
-    char_u *(pat[]) = { buf, buf2 };
-    if (gen_expand_wildcards(2, pat, &num_files, &files, EW_DIR) == OK) {
-      for (int i = 0; i < num_files; i++) {
-        push_path(search_path, xstrdup((char *)files[i]), false);
-        size_t after_size = STRLEN(files[i])+7;
-        char *after = xmallocz(after_size);
-        xstrlcpy(after, (char *)files[i], after_size);
-        xstrlcat(after, "/after", after_size);
-        if (os_isdir((char_u *)after)) {
-          kv_push(*after_path, after);
-        } else {
-          xfree(after);
-        }
-      }
-      FreeWild(num_files, files);
+  if (map_get(String, handle_T)(rtp_used, cstr_as_string(entry))) {
+    return;
+  }
+
+  if (!*entry) {
+    push_path(search_path, rtp_used, xstrdup(entry), after);
+  }
+
+  int num_files;
+  char_u **files;
+  char_u *(pat[]) = { (char_u *)entry };
+  if (gen_expand_wildcards(1, pat, &num_files, &files, EW_DIR) == OK) {
+    for (int i = 0; i < num_files; i++) {
+      push_path(search_path, rtp_used, xstrdup((char *)files[i]), after);
+    }
+    FreeWild(num_files, files);
+  }
+}
+
+static void expand_pack_entry(RuntimeSearchPath *search_path, Map(String, handle_T) *rtp_used,
+                              CharVec *after_path, char_u *pack_entry)
+{
+  static char buf[MAXPATHL];
+  char *(start_pat[]) = { "/pack/*/start/*", "/start/*" };  // NOLINT
+  for (int i = 0; i < 2; i++) {
+    if (!(STRLEN(pack_entry) + STRLEN(start_pat[i]) + 1 < MAXPATHL)) {
```

Comment:
nit: this negative logic (used in 3 places) could be more readable if the `!(... < ...)` is switched to `... >= ...`